_posts/2023-09-19-virtual-thread-1.adoc

```diff
@@ -0,0 +1,281 @@
+
+But what does it change?
+Blocking a virtual thread is, in general, very cheap! 
+There is a pinch of magic that makes virtual thread very appealing.
+When your code running on a virtual thread needs to execute an I/O operation, it uses a blocking API.
+So, the code waits for the result, as with the imperative model.
+However, since the JVM manages virtual threads, no underlying OS thread is blocked when they perform this blocking operation. 
+The state of the virtual thread is stored in the heap, and another virtual thread can be executed on the same Java platform (carrier) thread, exactly as in the reactive model.
+When the I/O operation completes, the virtual thread becomes executable again, and when a carrier thread is available, the state of the virtual thread is restored, and the execution continues.
+For the developer, all this mechanic is invisible! 
+You just write synchronous code, and it's executed like proper reactive code without blocking the OS thread.
+
+Your code runs on top of virtual threads, but under the hood, only a few carrier threads execute them. 
+
+To summarize, virtual threads are:
+
+    1. Lightweight - you can have a LOT of them
+    2. Cheap to create - no need to pool them anymore
+    3. Cheap to block - blocking a virtual thread does not block the underlying OS thread
+
+## How can you use virtual threads in Quarkus?
+
+Using virtual threads in Quarkus is straightforward.
+You only need to use the `@RunOnVirtualThread` annotation. 
+It indicates to Quarkus to invoke the annotated method on a virtual thread instead of a regular platform thread. 
+
+This new strategy extends the _smart dispatch_ explained in the https://quarkus.io/blog/resteasy-reactive-smart-dispatch/['to block or not to block' article]. In addition to the signature, Quarkus now looks for this specific annotation.
+If your JVM does not provide virtual thread support, it does fall back to platform threads. 
+
+Let's rewrite the same example using a virtual thread:
+
+[source, java]
+----
+@Path("/greetings")
+public class VirtualThreadApp {
+
+  @RestClient RemoteService service;
+
+  @GET
+  @RunOnVirtualThread
+  public String process() {
+    // Runs on a virtual thread because the
+    // method uses the @RunOnVirtualThread annotation.
+
+    // `service` is a rest client, it executes an I/O operation
+    var response = service.greetings(); // Blocking, but this time, it 
+                                        // does neither block the carrier thread
+                                        // nor the OS thread.
+                                        // Only the virtual thread is blocked.
+	return response.toUpperCase();
+  }
+
+}
+----
+
+It's the code from the first snippet (the imperative one), but its execution model is closer to the reactive one.
+This snippet works as described because the Quarkus REST client is virtual-thread-firendly.
```

Comment:
I think there should be more examples (not necessarily with code examples) of potencial usage of virtual threads... Like this, it could seem that only the HTTP layer is able to handle it.

Reply:
I've added a kafka example.